src/libraries/System.Private.CoreLib/src/System/Threading/RegisteredWaitHandle.Portable.cs

```diff
@@ -2,17 +2,123 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System.Diagnostics;
+using System.Runtime.Versioning;
+using Microsoft.Win32.SafeHandles;
 
 namespace System.Threading
 {
+    /// <summary>
+    /// An object representing the registration of a <see cref="WaitHandle"/> via <see cref="ThreadPool.RegisterWaitForSingleObject"/>.
+    /// </summary>
+    [UnsupportedOSPlatform("browser")]
     public sealed partial class RegisteredWaitHandle : MarshalByRefObject
     {
+        internal RegisteredWaitHandle(WaitHandle waitHandle, _ThreadPoolWaitOrTimerCallback callbackHelper,
+            int millisecondsTimeout, bool repeating)
+        {
+            Handle = waitHandle.SafeWaitHandle;
+            Callback = callbackHelper;
+            TimeoutDurationMs = millisecondsTimeout;
+            Repeating = repeating;
+            if (!IsInfiniteTimeout)
+            {
+                RestartTimeout();
+            }
+        }
+
+        private static AutoResetEvent? s_cachedEvent;
+
+        private static AutoResetEvent RentEvent()
+        {
+            AutoResetEvent? resetEvent = Interlocked.Exchange(ref s_cachedEvent, (AutoResetEvent?)null);
+            if (resetEvent == null)
+            {
+                resetEvent = new AutoResetEvent(false);
+            }
+            return resetEvent;
+        }
```

Comment:
Nit: this could be:
```C#
private static AutoResetEvent RentEvent() =>
    Interlocked.Exchange(ref s_cachedEvent, (AutoResetEvent?)null) ??
    new AutoResetEvent(false);
```